From 81890aca23ed59b44c806d0247b8b693faade080 Mon Sep 17 00:00:00 2001
From: "yanmei.yang" <yanmei.yang@amlogic.com>
Date: Thu, 29 Jul 2021 15:18:19 +0800
Subject: [PATCH] RDK: Kernel crash after exit STR mode [1/1]

PD#SWPL-56352

Problem:
Box enter into STR mode and exit it. kernel will crash during resume.

Solution:
check the lock before use it.

Verify:
check if kernel still crash or not during resume from STR mode.

Signed-off-by: yanmei.yang <yanmei.yang@amlogic.com>
Change-Id: Ie4ab0a179697c5a97af47f54c1797dd06c75bf81
---
 drivers/amlogic/drm/meson_hdmi.c | 15 ++++++++++-----
 1 file changed, 10 insertions(+), 5 deletions(-)

diff --git a/drivers/amlogic/drm/meson_hdmi.c b/drivers/amlogic/drm/meson_hdmi.c
index cad711fb98bf..8e5a26066fc6 100644
--- a/drivers/amlogic/drm/meson_hdmi.c
+++ b/drivers/amlogic/drm/meson_hdmi.c
@@ -473,12 +473,15 @@ static void meson_hdmitx_set_hdcp_result(int result)
 	struct drm_connector *connector = &am_hdmi_info.base.connector;
 	struct drm_modeset_lock *mode_lock =
 		&connector->dev->mode_config.connection_mutex;
+	bool locked_outer = drm_modeset_is_locked(mode_lock);
 
 	if (result == HDCP_AUTH_OK) {
 		am_hdmi_info.hdcp_state = HDCP_STATE_SUCCESS;
-		drm_modeset_lock(mode_lock, NULL);
+		if (!locked_outer)
+			drm_modeset_lock(mode_lock, NULL);
 		drm_hdcp_update_content_protection(connector, DRM_MODE_CONTENT_PROTECTION_ENABLED);
-		drm_modeset_unlock(mode_lock);
+		if (!locked_outer)
+			drm_modeset_unlock(mode_lock);
 		DRM_DEBUG("hdcp [%d] set result ok.\n", am_hdmi_info.hdcp_mode);
 	} else if (result == HDCP_AUTH_FAIL) {
 		am_hdmi_info.hdcp_state = HDCP_STATE_FAIL;
@@ -488,11 +491,13 @@ static void meson_hdmitx_set_hdcp_result(int result)
 		/*reset property value to DESIRED.*/
 		if (connector->state &&
 			connector->state->content_protection ==
-				DRM_MODE_CONTENT_PROTECTION_ENABLED) {
-			drm_modeset_lock(mode_lock, NULL);
+			DRM_MODE_CONTENT_PROTECTION_ENABLED) {
+			if (!locked_outer)
+				drm_modeset_lock(mode_lock, NULL);
 			drm_hdcp_update_content_protection(connector,
 				DRM_MODE_CONTENT_PROTECTION_DESIRED);
-			drm_modeset_unlock(mode_lock);
+			if (!locked_outer)
+				drm_modeset_unlock(mode_lock);
 		}
 	}
 }
-- 
2.29.2

